### ISSUE TITLE
Uncaught ReferenceError: exports is not defined

### ISSUE THREAD

**Issue by lukejagodzinski** [electron/electron-quick-start-typescript#27]
Thanks for updating this repository to use the newest version of Electron. I was trying to use it but I have a problem with React. When I try to import I get the `Uncaught ReferenceError: exports is not defined`. I've tried several solutions found on Stackoverflow and GitHub but nothing works. I've also tried `nodeIntegration: true` but it doesn't seem to work either. It exposes `window.require` function but it doesn't add `module.exports`. I don't know what's wrong with that.

**Comment by lewixlabs** [comment]
Same issue, no updates?

**Comment by Tenodec** [comment]
I  got the same  problem

**Comment by TTSKarlsson** [comment]
I have the same problem, and adding ```<script>var exports = {};</script>``` before any other script tags in index.html resolves this.

I wish we wouldn't have to do hacks like that, but I doubt it's a fault of this repo. Adding this script tag in the default index.html with a comment about why would probably help a lot of users.

**Comment by 5aitama** [comment]
Same problem, it can be resolved if you use webpack.
- First need to install it with `npm install webpack webpack-cli --save-dev` 
- Create a file named `webpack.config.js` *(in your project root folder)*
- Add this in your `webpack.config.js` :
```js
var path = require('path');

module.exports = {
	watch: true,
	target: 'node',
	mode: 'development',
	entry: `./dist/renderer.js`,
	output: {
		path: path.resolve(__dirname, 'dist'),
		filename: 'bundle.js'
	}
};
```
- In your `index.html` file, replace :
	```html 
	<script src="./dist/renderer.js"></script>
	```
	by this :
	```html
	<script src="./dist/bundle.js"></script>
	```
- Generate your `bundle.js` file with `npx webpack --config webpack.config.js`
- And now you can `npm start`

This work for me 😄

**Comment by htbkoo** [comment]
@lukejagodzinski 
Not sure if you still need this - I had encountered the same issue, and I might have found a workaround to fix it (which might not be the best practice but at least this had fixed my problem):

### TLDR
Set `webPreferences`.`nodeIntegration` to `true` when creating the `BrowserWindow`, e.g.
```
const mainWindow = new BrowserWindow({
    height: 600,
    webPreferences: {
      preload: path.join(__dirname, "preload.js"),
      nodeIntegration: true,
    },
    width: 800,
  });
```

### Details / my findings
I believe that since [`electron` v5.0.0](https://github.com/electron/electron/releases/tag/v5.0.0), `the default values of nodeIntegration and webviewTag are now false to improve security`

And just FYI, I had figured this out by looking at the [electron/simple-samples](https://github.com/electron/simple-samples/tree/master/activity-monitor) which are still working.
And I had found that the examples are using `electron` of older version (e.g. [v3.0.2](https://github.com/electron/simple-samples/blob/master/activity-monitor/package.json))
After that, I tried to play around with different versions of `electron` and the app would be broken after upgrading `electron` to v5.0.0.

### P.S
@vanessayuenn 
Thanks a lot for creating this starter repo! :-)
But I guess it would be great if you (or one of the maintainers) can share you thoughts on this / mention this somewhere in the documentation. Thank you!

**Comment by leontepe** [comment]
Is there any way to solve this without using Webpack? Using Webpack for Electron seems a little bit odd to me and I think there has to be some native way to fix this issue... any other ideas?

**Comment by Jamdan2** [comment]
I had this same problem, which only happened if my code included an `import` statement. @5aitama's solution worked for me.

**Comment by saostad** [comment]
in `main.ts` activate `nodeIntegration` like below:
```ts
webPreferences: {
      nodeIntegration: true,
      preload: path.join(__dirname, "preload.js"),
    },
```
in your `index.html`
replace: 
```html
<script src="./dist/renderer.js"></script>
```
with: 
```html
<script>
    require("./dist/renderer.js");
</script>
```

**Comment by jakekara** [comment]
@saostad's suggestion gives me an error that it violates the content security policy. Removing the CSP warns me that it's insecure.

Same issue with @TTSKarlsson's suggestion of inline script in index.html. However, I put the `var exports = {}` in a js file and added `<script src='set-exports.js'></script>` to index.html.

**Comment by duvet86** [comment]
> in `main.ts` activate `nodeIntegration` like below:
> 
> ```ts
> webPreferences: {
>       nodeIntegration: true,
>       preload: path.join(__dirname, "preload.js"),
>     },
> ```
> 
> in your `index.html`
> replace:
> 
> ```
> <script src="./dist/renderer.js"></script>
> ```
> 
> with:
> 
> ```
> <script>
>     require("./dist/renderer.js");
> </script>
> ```

@saostad, working but very unsafe: https://github.com/electron/electron/issues/9920#issuecomment-575839738

**Comment by ShlomiRex** [comment]
Still no fix?
I don't want to enable nodeIntegration...
Is there any way to fix this?

**Comment by Linch1** [comment]
@ShlomiRex check [this solution](https://stackoverflow.com/questions/54619111/typescript-electron-exports-is-not-defined) initially it didn't work also with `nodeintegration: true` but ( mysteriously ) after a few restart of electron started to work.

**Comment by akshaybabloo** [comment]
> in `main.ts` activate `nodeIntegration` like below:
> 
> ```ts
> webPreferences: {
>       nodeIntegration: true,
>       preload: path.join(__dirname, "preload.js"),
>     },
> ```
> 
> in your `index.html`
> replace:
> 
> ```
> <script src="./dist/renderer.js"></script>
> ```
> 
> with:
> 
> ```
> <script>
>     require("./dist/renderer.js");
> </script>
> ```

Also, set `contextIsolation` to `false`:

```ts
webPreferences: {
    nodeIntegration: true,
    contextIsolation: false,
    preload: path.join(__dirname, "preload.js"),
 }
```

I couldn't use it just by `nodeIntegration`.

**Comment by JoCat** [comment]
This is a dirty hack, but you can do it like this:

set the following settings in main.ts:
```ts
webPreferences: {
  preload: path.join(__dirname, "preload.js"),
},
```

in preload.ts use import like:
```ts
import { contextBridge, shell } from 'electron'

function openUrl(url: string) {
  shell.openExternal(url)
}

contextBridge.exposeInMainWorld("openUrl", openUrl);
```

in renderer.ts use like:
```ts
const _window = window as any; // Yes, it's dirty, but I have not yet figured out how to do it better (you can try to build an extended interface from Window, but so far I have not tried it)
_window.openUrl("https://example.com/");
```

**Comment by malept** [comment]
Here's my proposal for fixing this: https://github.com/electron/electron-quick-start-typescript/pull/152

**Comment by akii0008** [comment]
Any movement on this? @5aitama's solution does *not* work for me, saying it can't find module 'fs' in the electron node_module, oddly. Tried using an `exclude` regex to no avail.

**Comment by Kikketer** [comment]
> Here's my proposal for fixing this: #152

This is the fix

**Comment by malept** [comment]
> > Here's my proposal for fixing this: #152
> 
> This is the fix

It's not, see the [comments in the PR](https://github.com/electron/electron-quick-start-typescript/pull/152#pullrequestreview-914842997).
